src/mpi/init/finalize.c

```diff
@@ -120,6 +120,8 @@ int MPI_Finalize(void)
 
     MPIR_ERRTEST_INITIALIZED_ORDIE();
 
+    MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
with `direct-nm`, `MPI_Barrier` doesn't work -- why?

Replace with `PMI_Barrier` works. Is it a concern that `PMI_Barrier` maybe too slow?

Reply:
In a simple 2-node case, P0->P1 and P1->P0.  P0 exits as soon as it receives from P1 then proceed to tear down, breaking the injected message, leaving P1 hanging.

Reply:
OK, all we need to do is to set `MPIDI_OFI_global.max_buffered_send = 0` before the `MPI_Barrier`. We'll either need to move it to MPIR layer (suchas `MPIR_Process`) or have a setter function, e.g. `MPID_set_max_buffered_send(int thresh)`.